httpse.go

```diff
@@ -0,0 +1,143 @@
+package httpseverywhere
+
+import (
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/Yawning/obfs4/common/log"
+	"github.com/getlantern/golog"
+	iradix "github.com/hashicorp/go-immutable-radix"
+)
+
+// Rewrite changes an HTTP URL to rewrite.
+type Rewrite func(url *url.URL) (string, bool)
+
+type httpse struct {
+	log             golog.Logger
+	runs            int64
+	totalTime       int64
+	max             int64
+	maxHost         string
+	statM           sync.RWMutex
+	initOnce        sync.Once
+	wildcardTargets atomic.Value // *iradix.Tree
+	plainTargets    atomic.Value // map[string]*ruleset
+}
+
+// Default returns a lazily-initialized Rewrite using the default rules
+func Default() Rewrite {
+	h := newEmpty()
+	h.initAsync()
+	return h.rewrite
+}
+
+func newEmpty() *httpse {
+	h := &httpse{
+		log: golog.LoggerFor("httpse"),
+	}
+	h.wildcardTargets.Store(iradix.New())
+	h.plainTargets.Store(make(map[string]*ruleset))
+	return h
+}
+
+func (h *httpse) init() {
+	d := newDeserializer()
+	plain, wildcard, err := d.newRulesets()
+	if err != nil {
+		return
+	}
+	h.plainTargets.Store(plain)
+	h.wildcardTargets.Store(wildcard)
+}
+
+func (h *httpse) initAsync() {
+	h.initOnce.Do(func() {
+		go h.init()
+	})
+}
+
+func (h *httpse) rewrite(url *url.URL) (string, bool) {
+	if url.Scheme != "http" {
+		return "", false
+	}
+
+	start := time.Now()
+	if val, ok := h.plainTargets.Load().(map[string]*ruleset)[url.Host]; ok {
+		r, hit := h.rewriteWithRuleset(url, val)
+		h.addTiming(time.Now().Sub(start), url.String())
+		return r, hit
+	}
+	_, val, match := h.wildcardTargets.Load().(*iradix.Tree).Root().LongestPrefix([]byte(url.Host))
```

Comment:
As written, if there's a matching suffix rule we will only check `rewriteWithRuleset` for that suffix rule. If that returns false, we won't rewrite, even though there may be a matching prefix rule that would result in a rewrite.

To check both, you'd need to fall through on line 69 if `hit` is false.